### DISCUSSION TITLE
Not able to use pipeViewer due to 'extract_value' not defined error

### DISCUSSION THREAD

**Discussion by shubham-bhawalkar** [sparcians/map#446] [Q&A] [answered]
Hi guys, 

I see a bunch of open issues on the pipeViewer but wasn't obvious to me if any of them directly relate to what error I'm seeing in trying to use it. I'm using a Mac M1. Note that I'm able to build the core example just fine, with all the commands to run it as well. Just the pipeViewer is causing me pain. Would appreciate it if someone could help me out here :) [ Also please let me know if I should use some other way/forum to bring forth such issues]

I'm trying to build the pipeViewer tool after building Helios on the CoreExample on a fresh clone of map, and a fresh #build of Sparta. Not sure if I missed something, but in trying to follow the steps I'm running into this issue, here's the output after running the Argos.py command

`(sparta) Shubhams-MacBook-Pro:CoreModel shubham$ $HOME/map/helios/pipeViewer/pipe_view/argos.py \
       -l $HOME/map/sparta/example/CoreModel/cpu_layout.alf \
       -d my_pipeout`
       
`INFO: looking for  /Users/shubham/anaconda3/envs/sparta/bin/transactionsearch
transaction search initialized: you will be able to use search.
Using: /Users/shubham/anaconda3/envs/sparta/bin/transactionsearch
Traceback (most recent call last):
  File "/Users/shubham/map/helios/pipeViewer/pipe_view/argos.py", line 45, in <module>
    from pipe_view.model.workspace import Workspace  # noqa: E402
  File "/Users/shubham/anaconda3/envs/sparta/lib/python3.10/site-packages/pipe_view/model/workspace.py", line 15, in <module>
    from ..gui.layout_frame import Layout_Frame
  File "/Users/shubham/anaconda3/envs/sparta/lib/python3.10/site-packages/pipe_view/gui/layout_frame.py", line 17, in <module>
    from .layout_canvas import Layout_Canvas
  File "/Users/shubham/anaconda3/envs/sparta/lib/python3.10/site-packages/pipe_view/gui/layout_canvas.py", line 27, in <module>
    from .. import core
  File "pipe_view/core/src/core.pyx", line 11, in init pipe_view.core
NameError: name 'extract_value' is not defined`

**Comment by klingaard** [comment, marked as answer]
What version of cython do you have installed?  I had the same problem.  You'll need to make sure you have cython v0.29 installed.

**Reply by shubham-bhawalkar** [reply]
Fixed it, installed Cython 0.29.36 and got the cpu_layout.alf file copied over to core example directory and now it works :)

Thanks Knute!
<img width="791" alt="image" src="https://github.com/sparcians/map/assets/55960226/c81df7d3-0760-49be-b4dc-4585fea3daee">

**Reply by kennethpinzon** [reply]
Hi Knute, I have same issue. It persists even with cython 0.29.36 installed. Trying to understand where that extract_value definition is. I would appreciate your hints.
INFO: looking for  /home/xxx/mambaforge/envs/riscv_perf_model/bin/transactionsearch
transaction search initialized: you will be able to use search.
Using: /home/xxx/mambaforge/envs/riscv_perf_model/bin/transactionsearch
Traceback (most recent call last):
  File "/usr/local/tools/riscv/map/helios/pipeViewer/pipe_view/argos.py", line 45, in <module>
    from pipe_view.model.workspace import Workspace  # noqa: E402
  File "/home/xxx/mambaforge/envs/riscv_perf_model/lib/python3.10/site-packages/pipe_view/model/workspace.py", line 15, in <module>
    from ..gui.layout_frame import Layout_Frame
  File "/home/xxx/mambaforge/envs/riscv_perf_model/lib/python3.10/site-packages/pipe_view/gui/layout_frame.py", line 17, in <module>
    from .layout_canvas import Layout_Canvas
  File "/home/xxx/mambaforge/envs/riscv_perf_model/lib/python3.10/site-packages/pipe_view/gui/layout_canvas.py", line 27, in <module>
    from .. import core
  File "pipe_view/core/src/core.pyx", line 11, in init pipe_view.core
NameError: name 'extract_value' is not defined

**Reply by klingaard** [reply]
:thinking: Did you try Cython version 0.29.33?  Also, if you had this issue and then backed up `cython` and just tried building again, it might fail.  This is something I'm not happy with, but not all build files are kept in the `release` (for example) directory.

Try this, if you could (with your cython install of 0.29.36):
```
cd map/helios/pipeViewer
git clean -Xdf
cd ../release
make
```

**Reply by kennethpinzon** [reply]
Hi Knute, thanks for your quick response and sln. git clean -Xdf, did the work. I was able to start the pipeviewer and now I am trying to solve the some gtk issues.

I followed the instructions at https://github.com/riscv-software-src/riscv-perf-model/blob/master/README.md
I'm using mambaforge conda version 23.9 in a linux x86_64 machine. 

1. When a I tried to render the env to install the conda env for sparta, I couldnt render any yaml with the right dependencies
$./scripts/create_conda_env.sh sparta dev
output: 
```
+ cd ./scripts/..
+ ./scripts/render_recipe_for_platform.sh rendered_recipe.LWXquN
conda is /home/xxx/mambaforge/condabin/conda
jq is /home/xxx/mambaforge/envs/riscv_perf_model/bin/jq
yq is /home/xxx/mambaforge/envs/riscv_perf_model/bin/yq
Rendering recipe in /usr/local/tools/riscv/map/conda.recipe
+ conda render -m .ci_support/linux_64_.yaml --file rendered_recipe.LWXquN conda.recipe
::SUCCESS:: Rendered recipe written to 'rendered_recipe.LWXquN'
+ ./scripts/rendered2env_yaml.sh rendered_recipe.LWXquN dev
+ set -x
+ conda env create -f=environment.LWXquN.yml -n sparta
SpecNotFound: /usr/local/tools/riscv/map/environment.LWXquN.yml is not a valid yaml file
```

the environment.LWXquN.yml just contains:
```
channels:
  - conda-forge
dependencies:
```

I tried to make it work but the render_recipe_for_platform.sh using linux_64_.yaml didnt work. So I kept using the  riscv-perf-model env:

2. Use riscv-perf-model/conda/environment.yml as recipe.
and install the missing dependencies by hand for sparta/helios.
`conda install matplotlib-base numpy pillow pyyaml cython wxpython ipython
`
3. I was able to compile sparta, helios, and olympia.
With a warning after cmaking for helios:
```
CMake Warning at pipeViewer/transactionsearch/CMakeLists.txt:6 (add_executable):
  Cannot generate a safe runtime search path for target transactionsearch
  because files in some directories may conflict with libraries in implicit
  directories:

    runtime library [libsqlite3.so.0] in /home/xxx/mambaforge/envs/riscv_perf_model2/lib may be hidden by files in:
      /usr/lib/x86_64-linux-gnu

  Some of these libraries may not be found correctly.
```

4.  But once I tried to use the pipeviewer (after solving the Cython issue) there is a segment-fault (gtk2 installed)
$python $MAP_BASE/helios/pipeViewer/pipe_view/argos.py -d pipeout_1K -l ../layouts/small_core.alf

```
INFO: looking for  /home/xxx/mambaforge/envs/riscv_perf_model2/bin/transactionsearch
transaction search initialized: you will be able to use search.
Using: /home/xxx/mambaforge/envs/riscv_perf_model2/bin/transactionsearch
INFO:root:Constructing layout "../layouts/small_core.alf"
Constructing layout "../layouts/small_core.alf"
INFO:root:Layout loaded with 208 elements in 0.14158296585083008s
Layout loaded with 208 elements in 0.14158296585083008s

(argos.py:11173): GLib-GIO-CRITICAL **: 23:11:00.305: g_loadable_icon_load: assertion 'G_IS_LOADABLE_ICON (icon)' failed

(argos.py:11173): Gtk-WARNING **: 23:11:00.305: Error loading theme icon 'gtk-cancel' for stock: Failed to load icon

(argos.py:11173): Gtk-CRITICAL **: 23:11:00.305: render_fallback_image: assertion 'pixbuf != NULL' failed

(argos.py:11173): GLib-GObject-CRITICAL **: 23:11:00.305: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

(argos.py:11173): GLib-GIO-CRITICAL **: 23:11:00.516: g_loadable_icon_load: assertion 'G_IS_LOADABLE_ICON (icon)' failed

.
.
.
Segmentation fault

```

Now trying to solve these Gtk-CRITICAL assertions ... I do appreciate any hint.

**Reply by klingaard** [reply]
Appreciate the details!  Can you tell me the distro you're using?  For example, on Ubuntu:
```
lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

**Reply by kennethpinzon** [reply]
Thanks Knute for checking on this, using Debian 10:

```
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster
```

**Reply by klingaard** [reply]
Thanks!  I'll try to reproduce this error after the holidays.

**Reply by timsnyder** [reply]
Hi @kennethpinzon when you say

> render_recipe_for_platform.sh using linux_64_.yaml didnt work

Can you elaborate on what happened? This may need to become a separate ticket. I don't remember us doing any regression testing of hellos in the CI. ~~I don't know if the recipe for the conda env even includes the requirements for helios/argos.~~

Looking again, at
``` ################################################################################
#
#  BUILD MAP/HELIOS
#
################################################################################
pushd helios
mkdir -p release
pushd release
cmake -DCMAKE_BUILD_TYPE=Release \
      -DCMAKE_INSTALL_PREFIX:PATH="$PREFIX" \
      "${CMAKE_PLATFORM_FLAGS[@]}" \
      ..
cmake --build . -j "$CPU_COUNT" || cmake --build . -v
popd
popd
```
guess we build helios so the requirements should be in there. I'm still interested in what you ran into with the script.

**Reply by kennethpinzon** [reply]
Hi Tim, thanks for trying to help. 
Answering your questions:

1. Since I am not much familiar with conda I couldnt debug much what is going on. After running the conda render ```
command, the rendered_recipe is empty:
```
$ conda render -m .ci_support/linux_64_.yaml --file rendered_recipe.UCK2e8 conda.recipe
$ cat rendered_recipe.UCK2e8
$

```
2. After running
`$ cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX:PATH="$PREFIX" "${CMAKE_PLATFORM_FLAGS[@]}" ..
`
Output:

```
-- Looking for a built version of sparta in the source tree at /usr/local/tools/riscv/map2/helios/pipeViewer/../../sparta
-- HDF5 CXX compiler wrapper is unable to compile a minimal HDF5 program.
-- RapidJSON found. Headers: /home/xxx/mambaforge/envs/riscv_perf_model2/include
-- Configuring done
CMake Warning at pipeViewer/transactionsearch/CMakeLists.txt:6 (add_executable):
  Cannot generate a safe runtime search path for target transactionsearch
  because files in some directories may conflict with libraries in implicit
  directories:

    runtime library [libsqlite3.so.0] in /home/xxx/mambaforge/envs/riscv_perf_model2/lib may be hidden by files in:
      /usr/lib/x86_64-linux-gnu

  Some of these libraries may not be found correctly.


CMake Warning at pipeViewer/argos_dumper/CMakeLists.txt:9 (add_executable):
  Cannot generate a safe runtime search path for target Argos_dumper because
  files in some directories may conflict with libraries in implicit
  directories:

    runtime library [libsqlite3.so.0] in /home/xxx/mambaforge/envs/riscv_perf_model2/lib may be hidden by files in:
      /usr/lib/x86_64-linux-gnu

  Some of these libraries may not be found correctly.


CMake Warning at pipeViewer/argos_dumper/DatabaseDump/CMakeLists.txt:6 (add_executable):
  Cannot generate a safe runtime search path for target DBDumper because
  files in some directories may conflict with libraries in implicit
  directories:

    runtime library [libsqlite3.so.0] in /home/xxx/mambaforge/envs/riscv_perf_model2/lib may be hidden by files in:
      /usr/lib/x86_64-linux-gnu

  Some of these libraries may not be found correctly.

```
`$ cmake --build . -j "$CPU_COUNT" || cmake --build . -v
`
```
[ 14%] Building pipeViwer and its dependencies
[ 28%] Linking CXX executable transactionsearch
[ 42%] Linking CXX executable DBDumper
[ 57%] Linking CXX executable Argos_dumper
/home/xxx/mambaforge/envs/riscv_perf_model2/lib/python3.10/site-packages/setuptools/_distutils/extension.py:134: UserWarning: Unknown Extension options: 'pyrex_gdb'
  warnings.warn(msg)
/home/xxx/mambaforge/envs/riscv_perf_model2/lib/python3.10/site-packages/setuptools/__init__.py:80: _DeprecatedInstaller: setuptools.installer and fetch_build_eggs are deprecated.
!!

        ********************************************************************************
        Requirements should be satisfied by a PEP 517 installer.
        If you are using pip, you can try `pip install --use-pep517`.
        ********************************************************************************

!!
  dist.fetch_build_eggs(dist.setup_requires)
[ 71%] Built target DBDumper
running build_ext
[ 85%] Built target transactionsearch
[ 85%] Built target pipeView
[100%] Built target Argos_dumper
```

**Comment by kennethpinzon** [comment]
Just curious if someone could have replicated the issue or have any hints

**Reply by klingaard** [reply]
Specifically, which issue are you seeing?  If it's the missing `extract_value` problem, it's related to the version of `cython` -- it must be v0.29.

Are you having a different issue?

**Reply by kennethpinzon** [reply]
Thanks Knute. Once I tried to use the pipeviewer (after solving the Cython issue) there is a segment-fault (gtk2 installed)

```
$python $MAP_BASE/helios/pipeViewer/pipe_view/argos.py -d pipeout_1K -l ../layouts/small_core.alf
INFO: looking for  /home/xxx/mambaforge/envs/riscv_perf_model2/bin/transactionsearch
transaction search initialized: you will be able to use search.
Using: /home/xxx/mambaforge/envs/riscv_perf_model2/bin/transactionsearch
INFO:root:Constructing layout "../layouts/small_core.alf"
Constructing layout "../layouts/small_core.alf"
INFO:root:Layout loaded with 208 elements in 0.14158296585083008s
Layout loaded with 208 elements in 0.14158296585083008s

(argos.py:11173): GLib-GIO-CRITICAL **: 23:11:00.305: g_loadable_icon_load: assertion 'G_IS_LOADABLE_ICON (icon)' failed

(argos.py:11173): Gtk-WARNING **: 23:11:00.305: Error loading theme icon 'gtk-cancel' for stock: Failed to load icon

(argos.py:11173): Gtk-CRITICAL **: 23:11:00.305: render_fallback_image: assertion 'pixbuf != NULL' failed

(argos.py:11173): GLib-GObject-CRITICAL **: 23:11:00.305: g_object_ref: assertion 'G_IS_OBJECT (object)' failed

(argos.py:11173): GLib-GIO-CRITICAL **: 23:11:00.516: g_loadable_icon_load: assertion 'G_IS_LOADABLE_ICON (icon)' failed

.
.
.
Segmentation fault

```

**Reply by klingaard** [reply]
I've run into that seg fault before... trying to remember the issue.  Just to make sure, did you try these tricks?
```
cd map/helios/pipeViewer
git clean -Xdf
cd ../release
make
```

**Reply by kennethpinzon** [reply]
Yes, I applied those and solved my cython issue but now I get the seg fault

**Reply by knute-mips** [reply]
Remove this file and try again:
```
rm ~/.config/argos/config.json 
```

**Comment by kennethpinzon** [comment]
After deleting ~/.config/argos/config.json, same behaviour. And once I run
 `$python MAP_BASE/helios/pipeViewer/pipe_view/argos.py -d pipeout_1K -l ../layouts/small_core.alf`
the file ~/.config/argos/config.json gets created again

**Reply by knute-mips** [reply]
Correct, it will be created again, but hopefully the seg fault is no longer an issue.  If it still is seg faulting, please fire up a `gdb` on it and show me the stack.

**Comment by kennethpinzon** [comment]
Deleting the json file didn't work. I am getting tangled by a dailyjob-related project, expecting to get back in few days to dive in this issue. In the meantime, hints are welcome.

**Comment by klingaard** [comment]
This is now fixed in https://github.com/sparcians/map/pull/494
